src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -599,48 +970,41 @@ private void WriteElement(ref object o, CollectionMember collectionMember, Eleme
                 throw new InvalidOperationException(SR.Format(SR.XmlInternalError));
             }
 
-            if (choice != null && masterObject != null)
-            {
-                foreach (var name in choice.MemberIds)
-                {
-                    if (name == element.Name)
-                    {
-                        object choiceValue = Enum.Parse(choice.Mapping.TypeDesc.Type, name);
-                        SetOrAddValueToMember(masterObject, choiceValue, choice.MemberInfo);
+            member?.ChoiceSource?.Invoke(element.Name);
 
-                        break;
-                    }
-                }
-            }
-
-            if (collectionMember != null)
+            if (member?.ArraySource != null)
             {
-                collectionMember.Add(value);
+                member?.ArraySource(value);
             }
             else
             {
-                o = value;
+                member?.Source?.Invoke(value);
+                member?.CheckSpecifiedSource?.Invoke(true);
             }
+
+            return value;
         }
 
         private XmlSerializationReadCallback CreateXmlSerializationReadCallback(TypeMapping mapping)
         {
-            if (mapping is StructMapping)
+            if (mapping is StructMapping structMapping)
             {
-                return () => WriteStructMethod((StructMapping)mapping, mapping.TypeDesc.IsNullable, true, defaultNamespace: null);
+                return () => WriteStructMethod(structMapping, mapping.TypeDesc.IsNullable, true, defaultNamespace: null);
             }
-            else if (mapping is EnumMapping)
+            else if (mapping is EnumMapping enumMapping)
             {
-                return () => WriteEnumMethodSoap((EnumMapping)mapping);
+                return () => WriteEnumMethodSoap(enumMapping);
             }
-            else if (mapping is ArrayMapping)
-            {
-                return DummyReadArrayMethod;
-            }
-            else
+            else if (mapping is NullableMapping nullableMapping)
             {
-                throw new NotImplementedException();
+                return () => WriteNullableMethod(nullableMapping, false, null);
             }
+
+            return DummyReadArrayMethod;
+        }
+
+        private void NoopAction(object o)
```

Comment:
Make this method static so that it will only create a single instance of the implicit delegate for all instances.